pkg/storage/storage.go

```diff
@@ -80,69 +84,121 @@ func (p *storage) GetContainerMetrics(pods ...apitypes.NamespacedName) ([]api.Ti
 
 	timestamps := make([]api.TimeInfo, len(pods))
 	resMetrics := make([][]metrics.ContainerMetrics, len(pods))
-
-	for i, pod := range pods {
-		metricPoint, present := p.pods[pod]
+	for podIdx, pod := range pods {
+		contPoints, present := p.pods[pod]
 		if !present {
 			continue
 		}
 
-		contMetrics := make([]metrics.ContainerMetrics, len(metricPoint.Containers))
-		var earliestTS *time.Time
-		for i, contPoint := range metricPoint.Containers {
-			contMetrics[i] = metrics.ContainerMetrics{
+		contMetrics := make([]metrics.ContainerMetrics, len(contPoints))
+		var earliestTS time.Time
+		var contIdx int
+		for contName, contPoint := range contPoints {
+			prevContPoint := p.prevPods[pod][contName]
+			cpuUsage := cpuUsageOverTime(contPoint.MetricsPoint, prevContPoint.MetricsPoint)
+			contMetrics[contIdx] = metrics.ContainerMetrics{
 				Name: contPoint.Name,
 				Usage: corev1.ResourceList{
-					corev1.ResourceName(corev1.ResourceCPU):    contPoint.CpuUsage,
+					corev1.ResourceName(corev1.ResourceCPU):    cpuUsage,
 					corev1.ResourceName(corev1.ResourceMemory): contPoint.MemoryUsage,
 				},
 			}
-			if earliestTS == nil || earliestTS.After(contPoint.Timestamp) {
-				ts := contPoint.Timestamp // copy to avoid loop iteration variable issues
-				earliestTS = &ts
+			if earliestTS.IsZero() || earliestTS.After(contPoint.Timestamp) {
+				earliestTS = contPoint.Timestamp
 			}
+			contIdx++
 		}
-		if earliestTS == nil {
-			// we had no containers
-			earliestTS = &time.Time{}
-		}
-		timestamps[i] = api.TimeInfo{
-			Timestamp: *earliestTS,
+		resMetrics[podIdx] = contMetrics
+		timestamps[podIdx] = api.TimeInfo{
+			Timestamp: earliestTS,
 			Window:    kubernetesCadvisorWindow,
 		}
-		resMetrics[i] = contMetrics
 	}
 	return timestamps, resMetrics, nil
 }
 
 func (p *storage) Store(batch *MetricsBatch) {
-	newNodes := make(map[string]NodeMetricsPoint, len(batch.Nodes))
-	var nodeCount, containerCount int
-	for _, nodePoint := range batch.Nodes {
+	p.storeNodeMetrics(batch.Nodes)
+	p.storePodMetrics(batch.Pods)
+}
+
+func (p *storage) storeNodeMetrics(nodes []NodeMetricsPoint) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	newNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	prevNodes := make(map[string]NodeMetricsPoint, len(nodes))
+	var nodeCount int
+	for _, nodePoint := range nodes {
 		if _, exists := newNodes[nodePoint.Name]; exists {
 			klog.Errorf("duplicate node %s received", nodePoint.Name)
 			continue
 		}
-		nodeCount += 1
 		newNodes[nodePoint.Name] = nodePoint
+		nodeCount++
+
+		if nodePoint.Timestamp.After(p.nodes[nodePoint.Name].Timestamp) {
+			prevNodes[nodePoint.Name] = p.nodes[nodePoint.Name]
+		} else {
+			prevNodes[nodePoint.Name] = p.prevNodes[nodePoint.Name]
+		}
+		nodeCount++
```

Comment:
About pointsStored metric, it was meant for completeness SLI of metric data stored, For example if there is one node in cluster we should have 1 metric point stored.

In your PR technically we double number of points stored physically (we store current and previous metric value), but logically it's still one as when storage is queries we can only answer if we have both points (both are required to calculate cpu usage). 

I think we should interpret this metric as less about points stored in memory, but about how many nodes we can return metric points when queried.
Based on that metric should be equal number of nodes that are available in both `newNodes` and `prevNodes`

Reply:
:+1: makes sense, I had indeed misunderstood the purpose of this metric.

Reply:
:+1: make sense, I had indeed misunderstood the original purpose of this metric